src/transformStateWithClones.js

```diff
@@ -7,7 +7,24 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    if (action.type === `addProperties`) {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === `removeProperties`) {
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    } else if (action.type === `clear`) {
+      currentState = {};
+    }
+
+    stateHistory.push({ ...currentState });
```

Comment:
You are correctly creating a new copy of the `currentState` before pushing it to `stateHistory`. This ensures that you don't mutate the original `state` object, adhering to the checklist.